fsm/types.go

```diff
@@ -0,0 +1,151 @@
+package fsm
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-statestore"
+)
+
+// EventName is the name of an event
+type EventName interface{}
+
+// Context provides access to the statemachine inside of a state handler
+type Context interface {
+	// Context returns the golang context for this context
+	Context() context.Context
+
+	// Event initiates a state transition with the named event.
+	//
+	// The call takes a variable number of arguments that will be passed to the
+	// callback, if defined.
+	//
+	// It will return nil if the event is one of these errors:
+	//
+	// - event X does not exist
+	//
+	// - arguments don't match expected transition
+	Event(event EventName, args ...interface{}) error
+}
+
+// ApplyTransitionFunc modifies the state further in addition
+// to modifying the state key. It the signature
+// func applyTransition<StateType, T extends any[]>(s stateType, args ...T)
+// and then an event can be dispatched on context or group
+// with the form .Event(Name, args ...T)
+type ApplyTransitionFunc interface{}
```

Comment:
I'm not familiar with this naming convention...  this would have been be clearer to if it was named something like `StateMutatorFunc` or `EventCallbackFunc` and `WithMutator()`, etc.  But if it's a well-known naming convention I'm cool with it.

Reply:
no I just made it up will change :)

Reply:
I'm leaning towards "actions" cause that is the actual FSM terminolgy for things that happen along with an event